src/Compilers/CSharp/Portable/Lowering/LambdaRewriter/LambdaRewriter.Analysis.Tree.cs

```diff
@@ -0,0 +1,505 @@
+                        var scope = _currentScope;
+                        var closure = _currentClosure;
+                        while (closure != null && symbol.ContainingSymbol != closure.OriginalMethodSymbol)
+                        {
+                            closure.CapturedVariables.Add(symbol);
+
+                            // Also mark captured in enclosing scopes
+                            while (scope.ContainingClosure == closure)
+                            {
+                                scope = scope.Parent;
+                            }
+                            closure = scope.ContainingClosure;
+                        }
+
+                        // Also record where the captured variable lives
+
+                        // No need to record where local functions live: that was recorded
+                        // in the Closures list in each scope
+                        if (symbol.Kind == SymbolKind.Method)
+                        {
+                            return;
+                        }
+
+                        // The 'this' parameter isn't declared in method scope
+                        if (symbol is ParameterSymbol param && param.IsThis)
+                        {
+                            return;
+                        }
+
+                        if (_localToScope.TryGetValue(symbol, out var declScope))
+                        {
+                            declScope.DeclaredCapturedVariables.Add(symbol);
+                        }
+                        else
+                        {
+                            // Parameters and locals from expression tree lambdas
+                            // don't get recorded
+                            Debug.Assert(_inExpressionTree);
+                        }
+                    }
+                }
+
+                /// <summary>
+                /// Create a new nested scope under the current scope, or reuse the current
+                /// scope if there's no change in the bound node for the nested scope.
+                /// Records the given locals as declared in the aforementioned scope.
+                /// </summary>
+                private Scope CreateOrReuseScope<TSymbol>(BoundNode node, ImmutableArray<TSymbol> locals)
+                    where TSymbol : Symbol
+                {
+                    // We should never create a new scope with the same bound
+                    // node. We can get into this situation for methods and
+                    // closures where a new scope is created to add parameters
+                    // and a new scope would be created for the method block,
+                    // despite the fact that they should be the same scope.
+                    var scope = _currentScope.BoundNode == node
+                        ? _currentScope
+                        : CreateNestedScope(node, _currentScope, _currentClosure);
+                    DeclareLocals(scope, locals);
```

Comment:
"OrReuse" part seems to cause a lot of additional handling. Perhaps it is easier to - 
* create the parameter scope associated with top node (or with null) and put parameters in it, if there are parameters, make the root scope its child.
* follow a simpler push/pop routine while building the nodes. In particular - no need to remember `oldScope` since it is always the parent of the pushed one. 
* when all is done merge the parameter scope and the root scope.

I think it might get a lot simpler while producing the same scope tree.